arch-design-questions.md

```diff
@@ -0,0 +1,93 @@
+# Client design
+
+## Principles
+
+- There are two top level clients - `ServiceBusReceiverClient` and `ServiceBusSenderClient`.
+- Clients are [single-use](#single-use-clients) - they are created with a receive mode, session ID (if applicable), queue/topic-subscription
+  baked in and cannot be changed.
+- The user can choose between two different methods of message delivery - push (via registered message handlers)
+  or push (via an iterator/reactor pattern or .receiveBatch()-style)
```

Comment:
"push or push"  (I assume the latter should be pull?  However the iterator in python is basically part of our push pattern, where pull would be "fetch_next" (which is effectively receiveBatch as far as I understand; e.g. imperatively fetch N events))

Reply:
Hm...perhaps push and pull are bad terms for this (although we've been using that internally).

When I think of 'push' I think of 'event handlers/callbacks'. So the "other" things are pull - perhaps our terms are bad there.

Reply:
@jsquire  - I picked up this terminology from our time on eventhubs. What do you call the iterator based approach in EventHubs? Push or pull? And do you mind telling us the reasoning? It'll help when we align.

Reply:
I've been calling it "iterator" since it seems to be the closest thing to a conceptual term that is understood and exists in some form across languages.   For .NET specifically, it's an "enumerator" or "async enumerator".

Reply:
I'm with you on the "handlers/callbacks" being the "other" side of that coin.   That's what I've been using while trying like mad to avoid using the .NET "event handler" term due to the overloaded context _(and my general dislike for the pattern over higher order functions, but I digress....)_

Reply:
So @KieranBrantnerMagee  perhaps we have three things:
* Push (event handler)
* Pull (receiveBatch/receive(x num of messages)
* Iterator

Reply:
I try to avoid those terms, as they seem to mean different things to different people.   If I really need to, I'll refer to it as "pulling values out of the iterator"

Reply:
Though, I'd also mention that you can have an iterator that exposes single values or one that returns batches.   It's more about the model for consuming it where `receive` or `receiveBatch` gets wrapped in the language construct (`foreach` in .NET's case) rather than someone writing a manual loop around it.

The iterator advantage is that it provides a transparent but explicit scope.   When the `foreach` is active, the scope is live and when that loop goes away, there's an implicit call to dispose, so you have a deterministic point in time where you know the caller has signaled the intent for "I'm done with this thing"

Reply:
I've updated the text to be a bit more granular. In particular Java's Flux pattern is just far more flexible and ends up living in two categories (eternal iteration vs finite receive).